### ISSUE TITLE
Use Ctrl + e quick conversion editor to search for card dead and view the log unhandled event loop exception

### ISSUE THREAD

**Issue by lyh061619** [eclipse-platform/eclipse.platform.ui#280]
![20220822114918](https://user-images.githubusercontent.com/15118571/185834902-3f0a0b0e-612e-4384-b321-6a4acfc70aa5.png)
**Describe the bug**
Use Ctrl + e quick conversion editor to search for card dead and view the log unhandled event loop exception
**To Reproduce**
A standalone SWT snippet to reproduce the behavior (preferable)
(or) 
Steps to reproduce the behavior.
Use Ctrl + e quick conversion editor to search 
**Expected behavior**
After the IDE is used for a period of time, the problem occurs and the IDE cannot be operated. The IDE is forced to be closed
**Screenshots**
Please see the attachment
**Environment:**
1. Select the platform(s) on which the behavior is seen:
- - [] All OS
- - [x] Windows
- - [ ] Linux
- - [ ] macOS 

2. Additional OS info (e.g. OS version, Linux Desktop, etc)

3. JRE/JDK version
 BellSoft jdk17.0.4
**Version since**
 Eclipse or SWT version since when the behavior is seen [e.g. 4.23]
4.24
[.log](https://github.com/eclipse-platform/eclipse.platform.swt/files/9391084/default.log)

**Workaround (or) Additional context**
Any known workarounds for the problem?

**Comment by Phillipus** [comment]
I guess this means using Ctrl-E to switch open editors. I tried this several times and found a lot of error log entries.  I think that when you type in the search field a bold font is used and not disposed?

```
java.lang.Error: SWT Resource was not properly disposed
	at org.eclipse.swt.graphics.Resource.initNonDisposeTracking(Resource.java:172)
	at org.eclipse.swt.graphics.Resource.<init>(Resource.java:120)
	at org.eclipse.swt.graphics.Font.<init>(Font.java:111)
	at org.eclipse.jface.viewers.BoldStylerProvider.getBoldFont(BoldStylerProvider.java:89)
	at org.eclipse.jface.viewers.BoldStylerProvider$1.applyStyles(BoldStylerProvider.java:65)
	at org.eclipse.jface.viewers.StyledString.createStyleRange(StyledString.java:475)
	at org.eclipse.jface.viewers.StyledString.getStyleRanges(StyledString.java:439)
	at org.eclipse.ui.internal.WorkbookEditorsHandler$1.update(WorkbookEditorsHandler.java:149)
	at org.eclipse.jface.viewers.ViewerColumn.refresh(ViewerColumn.java:144)
	at org.eclipse.jface.viewers.AbstractTableViewer.doUpdateItem(AbstractTableViewer.java:396)
	at org.eclipse.jface.viewers.StructuredViewer$UpdateItemSafeRunnable.run(StructuredViewer.java:427)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.viewers.StructuredViewer.updateItem(StructuredViewer.java:2107)
	at org.eclipse.jface.viewers.AbstractTableViewer.createItem(AbstractTableViewer.java:288)
	at org.eclipse.jface.viewers.AbstractTableViewer.internalRefreshAll(AbstractTableViewer.java:726)
	at org.eclipse.jface.viewers.AbstractTableViewer.internalRefresh(AbstractTableViewer.java:618)
	at org.eclipse.jface.viewers.AbstractTableViewer.internalRefresh(AbstractTableViewer.java:610)
	at org.eclipse.jface.viewers.StructuredViewer.lambda$2(StructuredViewer.java:1459)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1398)
	at org.eclipse.jface.viewers.StructuredViewer.preservingSelection(StructuredViewer.java:1359)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1459)
	at org.eclipse.jface.viewers.ColumnViewer.refresh(ColumnViewer.java:526)
	at org.eclipse.jface.viewers.StructuredViewer.refresh(StructuredViewer.java:1418)
	at org.eclipse.ui.internal.FilteredTableBaseHandler.lambda$1(FilteredTableBaseHandler.java:460)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:183)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4251)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1090)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1071)
	at org.eclipse.swt.widgets.Text.wmCommandChild(Text.java:3106)
	at org.eclipse.swt.widgets.Control.WM_COMMAND(Control.java:4896)
	at org.eclipse.swt.widgets.Control.windowProc(Control.java:4748)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:5036)
	at org.eclipse.swt.internal.win32.OS.CallWindowProc(Native Method)
	at org.eclipse.swt.widgets.Text.callWindowProc(Text.java:295)
	at org.eclipse.swt.widgets.Control.windowProc(Control.java:4846)
	at org.eclipse.swt.widgets.Text.windowProc(Text.java:2687)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:5028)
	at org.eclipse.swt.internal.win32.OS.DispatchMessage(Native Method)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3643)
	at org.eclipse.ui.internal.FilteredTableBaseHandler.openDialog(FilteredTableBaseHandler.java:306)
	at org.eclipse.ui.internal.FilteredTableBaseHandler.execute(FilteredTableBaseHandler.java:131)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:283)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:97)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:58)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:317)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:251)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:173)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:156)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:488)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:485)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:213)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.executeCommand(KeyBindingDispatcher.java:308)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.press(KeyBindingDispatcher.java:580)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.processKeyEvent(KeyBindingDispatcher.java:647)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.filterKeySequenceBindings(KeyBindingDispatcher.java:439)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher$KeyDownFilter.handleEvent(KeyBindingDispatcher.java:96)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.filterEvent(Display.java:1262)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1065)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1090)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1075)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1117)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1113)
	at org.eclipse.swt.widgets.Widget.wmChar(Widget.java:1542)
	at org.eclipse.swt.widgets.Control.WM_CHAR(Control.java:4867)
	at org.eclipse.swt.widgets.Canvas.WM_CHAR(Canvas.java:345)
	at org.eclipse.swt.widgets.Control.windowProc(Control.java:4745)
	at org.eclipse.swt.widgets.Canvas.windowProc(Canvas.java:340)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:5028)
	at org.eclipse.swt.internal.win32.OS.DispatchMessage(Native Method)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3643)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:643)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:550)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
```

I guess this could lead to no more SWT handles over time.

**Comment by Phillipus** [comment]
I think the problem starts at `org.eclipse.ui.internal.WorkbookEditorsHandler$1.update(WorkbookEditorsHandler.java:149)`

```java
StyledString ss = ssh.highlight(text, pattern,
    new BoldStylerProvider(WorkbookEditorsHandler.this.getFont(false, true))
       .getBoldStyler());
```

A new instance of `BoldStylerProvider` is created each time and never disposed. So 100s of bold fonts are created and resources not disposed and ultimately no more handles.

**Comment by mickaelistria** [comment]
Transfered to eclipse.platform.text as @Phillipus identified this is not a generic StyledText issue but instead an editor one.
@lyh061619 would you be interested in trying to submit a fix according to last comments?

**Comment by iloveeclipse** [comment]
@trancexpress : is this a regression in 4.24? If yes, *and* if the fix would be trivial,  could you please do this for RC1?

**Comment by Phillipus** [comment]
This one is tricky. I tried this:

```java
BoldStylerProvider boldStyleProvider = new BoldStylerProvider(WorkbookEditorsHandler.this.getFont(false, true));
StyledString ss = ssh.highlight(text, pattern, boldStyleProvider.getBoldStyler());
cell.setStyleRanges(ss.getStyleRanges());
boldStyleProvider.dispose();
```

It fixes the resource leak but bold font is not applied in the search field of the Editor switcher.

**Comment by Phillipus** [comment]
> is this a regression in 4.24?

It's in 4.23

**Comment by trancexpress** [comment]
@Phillipus : What are the exact reproduction steps for the leaked fonts? I don't see logged errors about undisposed resources when I try to use Ctrl+E with 4.25 on Linux (typing, selecting, etc.).

**Comment by Phillipus** [comment]
1. Open several editors
2. Ctrl-E to open the Editor switcher
3. Type in the search field a character so that the table of open editors highlights in bold

I have a fix incoming....

**Comment by laeubi** [comment]
@trancexpress please note that you maybe only see this on a GC-run that is happen somewhere sometimes and not related to a particular action!

**Comment by Phillipus** [comment]
Try this:

```java
	protected void setLabelProvider(final TableViewerColumn tableViewerColumn) {

		tableViewerColumn.setLabelProvider(new StyledCellLabelProvider() {
		    BoldStylerProvider boldStyleProvider = new BoldStylerProvider(WorkbookEditorsHandler.this.getFont(false, true));

			/* called once for each element in the table */
			@Override
			public void update(ViewerCell cell) {
				Object element = cell.getElement();
				if (element instanceof WorkbenchPartReference) {
					WorkbenchPartReference ref = (WorkbenchPartReference) element;
					String text = getWorkbenchPartReferenceText(ref);
					cell.setText(text);
					cell.setImage(ref.getTitleImage());

					SearchPattern matcher = WorkbookEditorsHandler.this.getMatcher();
					if (matcher == null) {
						cell.setStyleRanges(null);
					} else {
						String pattern = matcher.getPattern();
						StyledStringHighlighter ssh = new StyledStringHighlighter();
						StyledString ss = ssh.highlight(text, pattern, boldStyleProvider.getBoldStyler());
						cell.setStyleRanges(ss.getStyleRanges());
					}

					cell.getControl().redraw();
				}
			}

			@Override
			public String getToolTipText(Object element) {
				if (element instanceof WorkbenchPartReference) {
					WorkbenchPartReference ref = (WorkbenchPartReference) element;
					return ref.getTitleToolTip();
				}
				return super.getToolTipText(element);
			}

			@Override
			public void dispose() {
			    super.dispose();
			    boldStyleProvider.dispose();
			}
		});

		ColumnViewerToolTipSupport.enableFor(tableViewerColumn.getViewer());
	}
```

**Comment by Phillipus** [comment]
The above fix works. I create one `boldStyleProvider` that can be re-used and is disposed when the label provider is disposed. WDYT?

**Comment by Phillipus** [comment]
This is in `platform.ui` not `platform.text`

**Comment by laeubi** [comment]
> WDYT

Create a PR so we can see it passes the validation :+1:  
That would also make it easier to discuss possible code changes.

**Comment by Phillipus** [comment]
Should this issue be moved to `platform.ui` so I can carry on with PR there?

**Comment by Phillipus** [comment]
Was introduced here:

https://github.com/eclipse-platform/eclipse.platform.ui/commit/c5b24aa4ac0e40663a2741eddd31fa71ffaf1992

**Comment by laeubi** [comment]
> Should this issue be moved to `platform.ui` so I can carry on with PR there?

Done.

**Comment by Phillipus** [comment]
> Create a PR so we can see it passes the validation 👍

https://github.com/eclipse-platform/eclipse.platform.ui/pull/281

**Comment by Phillipus** [comment]
Would be nice to get this in for 4.25. Although it is a bit obscure, and is only triggered if you switch editors a lot with Ctrl-E _and_ type in the popup's search field a lot, and have Eclipse open for a long time... but the OP's screenshot isn't a good look. 😄

**Comment by lyh061619** [comment]
> Would be nice to get this in for 4.25. Although it is a bit obscure, and is only triggered if you switch editors a lot with Ctrl-E _and_ type in the popup's search field a lot, and have Eclipse open for a long time... but the OP's screenshot isn't a good look. 😄

Yes, I opened eclipse for a long time, frequently used Ctrl + e to switch and input keywords in the pop-up to search for the triggered bug. The above screenshot shows that there is no operation

**Comment by Phillipus** [comment]
> Yes, I opened eclipse for a long time, frequently used Ctrl + e to switch and input keywords in the pop-up to search for the triggered bug. The above screenshot shows that there is no operation

If you look in Eclipse's older error logs you'll probably see the error log entries showing the cause.

**Comment by lyh061619** [comment]
> > Yes, I opened eclipse for a long time, frequently used Ctrl + e to switch and input keywords in the pop-up to search for the triggered bug. The above screenshot shows that there is no operation
> 
> If you look in Eclipse's error log you'll probably see the error log entries showing the cause. It's always a good idea to check the error log to find a problem.

In the attachment, you can view the runtime exception log and view the location when it occurs! MESSAGE Unhandled event loop exception

! STACK 0

org.eclipse.swt. Swerror: no more handles, followed by eclipse component stack exception information

**Comment by Phillipus** [comment]
I meant the underlying cause of the resource leak, probably in an older eclipse log file.

```
java.lang.Error: SWT Resource was not properly disposed
	at org.eclipse.swt.graphics.Resource.initNonDisposeTracking(Resource.java:172)
	at org.eclipse.swt.graphics.Resource.<init>(Resource.java:120)
	at org.eclipse.swt.graphics.Font.<init>(Font.java:138)
	at org.eclipse.jface.viewers.BoldStylerProvider.getBoldFont(BoldStylerProvider.java:89)
```

**Comment by lyh061619** [comment]
> I meant the underlying cause of the resource leak, probably in an older eclipse log file.
> 
> ```
> java.lang.Error: SWT Resource was not properly disposed
> 	at org.eclipse.swt.graphics.Resource.initNonDisposeTracking(Resource.java:172)
> 	at org.eclipse.swt.graphics.Resource.<init>(Resource.java:120)
> 	at org.eclipse.swt.graphics.Font.<init>(Font.java:138)
> 	at org.eclipse.jface.viewers.BoldStylerProvider.getBoldFont(BoldStylerProvider.java:89)
> ```
The complete stack is in the new upload attachment,

ENTRY org.eclipse.ui.ide 4 4 2022-08-22 11:51:55.375
!MESSAGE Not properly disposed SWT resource
!STACK 0
java.lang.Error: SWT Resource was not properly disposed
	at org.eclipse.swt.graphics.Resource.initNonDisposeTracking(Resource.java:172)
	at org.eclipse.swt.graphics.Resource.<init>(Resource.java:120)
	at org.eclipse.swt.graphics.Font.<init>(Font.java:111)
	at org.eclipse.jface.viewers.BoldStylerProvider.getBoldFont(BoldStylerProvider.java:89)
	at org.eclipse.jface.viewers.BoldStylerProvider$1.applyStyles(BoldStylerProvider.java:65)
	at org.eclipse.jface.viewers.StyledString.createStyleRange(StyledString.java:475)
	at org.eclipse.jface.viewers.StyledString.getStyleRanges(StyledString.java:439)
	at org.jkiss.dbeaver.ui.SearchCellLabelProvider.update(SearchCellLabelProvider.java:70)
	at org.eclipse.jface.viewers.ViewerColumn.refresh(ViewerColumn.java:144)
	at org.eclipse.jface.viewers.AbstractTableViewer.doUpdateItem(AbstractTableViewer.java:396)
	at org.eclipse.jface.viewers.StructuredViewer$UpdateItemSafeRunnable.run(StructuredViewer.java:427)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
[.bak_8.log](https://github.com/eclipse-platform/eclipse.platform.ui/files/9394697/default.bak_8.log)

**Comment by lyh061619** [comment]
In multiple log file backups, the exception information is recorded

**Comment by Phillipus** [comment]
Yep, that's it. Anyway, I've created a fix in #281 and this will probably make it into Eclipse 4.26.

A workaround is to not use the search field so much, or close and re-start Eclipse every so often.
